### DISCUSSION TITLE
Servo motor + LED control with video capture

### DISCUSSION THREAD

**Discussion by KeertanaV** [bonsai-rx/bonsai#1088] [Q&A]
Hi all, I am rather new to bonsai and am trying to set up a system for my behavior tracking. Here I need the camera to start recording after 2mins for a period of 5 mins. In this 5 mins window, I need a servo to move 5 times from 0 to 40 degrees (at 2min30secs) and then a blue LED switches on for a period of 3 secs (at 2min 50secs). I have the camera and LED in a common workflow but am struggling with incorporating the servo motor into this. The timer controlling the servo motor is regulating the periods of the sweep angle because I need around 5 sweeps in 3secs. But I do not know how to give a delay for this timer as I need the servo to start only after 30secs of recording. Please could you help me with this? Below are the two workflows (1 - video capture with LED control,  2 - servo control)
1) ![video+LED](https://user-images.githubusercontent.com/107936031/196225831-07d0d8d6-6fdc-4475-a6ae-a4f8ee6926c8.PNG)

2) ![Servo](https://user-images.githubusercontent.com/107936031/196225853-b7f7ce85-801c-483e-a580-f486be77c78b.PNG)

**Comment by bruno-f-cruz** [comment]
Heya and welcome to Bonsai. 

I have attached a possible solution (a lot of room for improvement, especially on how the timers relate to each other) that shows how you can delay the start of your servo logic. Basically, encapsulate everything inside a [`SelectMany`](https://bonsai-rx.org/docs/api/Bonsai.Reactive.SelectMany.html) and trigger the node with a `Timer`.

Hope it helps 
B
[delayservo.zip](https://github.com/bonsai-rx/bonsai/files/9802618/delayservo.zip)

**Reply by KeertanaV** [reply]
Hi Bruno,

Thanks a lot for this! However, I am not able to open the file for some reason. Could you please share it again?

Thanks,
Keertana

**Reply by bruno-f-cruz** [reply]
I have tried it on another computer and it seems to be working. What is the error you get when trying to open it? Make sure you unzip the file (e.g. using [7zip](https://www.7-zip.org/)).

**Reply by KeertanaV** [reply]
Yes, I did unzip it but I am getting an OpenError like this:
![image](https://user-images.githubusercontent.com/107936031/196371396-44f9bc14-82fa-4892-a7fb-1d36aa9b3b0d.png)

**Reply by bruno-f-cruz** [reply]
Can you try to open it with the latest bonsai version?

**Reply by KeertanaV** [reply]
Hi Bruno, yes it did open in the latest version. Thanks.
However, there are some issues I am facing:
1) the servo is starting right away after around 4secs without following the delay
2) Also, the servo continues without a stop and I tried using the RepeatCount to define the number of sweeps but that did not work.
3) How can I modify the servo angle if required?

**Reply by KeertanaV** [reply]
And, also once the servo starts the camera and LED does not start. So it is something with the timers but I am confused about it.

**Reply by bruno-f-cruz** [reply]
Can you send the exact workflow you are using?

**Reply by KeertanaV** [reply]
[delayservo.zip](https://github.com/bonsai-rx/bonsai/files/9809560/delayservo.zip)

I added the workflow output after the videowriter. The camera and LED are now okay but the servo starts after 4-5secs and and goes on and any change in the due time or period for the timer is not making a difference.

**Reply by bruno-f-cruz** [reply]
From the workflow you sent me, the Servo will sweep for the first time after 20secs (or 10 secs from the start of the video). It will also repeat forever. You can add a "TakeUntil" controlled by a timer after the ServoSweep to make it stop after X amount of time

**Reply by KeertanaV** [reply]
Yes, but for some reason that is not happening. The servo starts irrespective after around 4secs every time and does not stop with the take until. The timer values are correct but I do not know what else is triggering the servo.

**Reply by KeertanaV** [reply]
And also when I run my old workflow for just the servo, it works fine.

**Reply by chiyu1203** [reply]
Hello, I found the use of Timer in this delayservo workflow interesting but it was somehow different from my have learned before at Bonsai tutorial <https://bonsai-rx.org/docs/tutorials/state-machines.html> so I hope you can share with me more details about the parameters used in the two Timers (one outside the SelectMany, the other one inside the SelectMany). 

At the Bonsai tutorial, I learned that we can use the property Period in the Timer, to trigger downstream module periodically. The servo motor will be turned on and off based on the Period (at the same speed). However, in your workflow, it seems that the Timer module in the SelectMany control the on and off set based on both property Period and DueTime respectively (and because it takes more time to execute the property Period than the DueTime, the time needs to be set 0.1 sec longer, as to 0.4 v.s 0.3 sec?!) In addition, the properties of the Timer module outside the SelectMany have two function. One (DueTime) serves something similar to DelayScription's function and the other one (Period) serves as Repeat's function?! However, the Period property here needs to match with how long it takes to run one SelectMany to makes this workflow repeatedly indefinitely?! What's the benefit of using this Timer function as opposed to using DelayScription and Repeat module?

**Reply by bruno-f-cruz** [reply]
Like any other programming language, in Bonsai you can usually solve a problem in many different ways.
Here, I leverage the fact that the property `Period` only starts "counting" once the `DueTime` property elapses. So having those 2 properties set and having a `Take(2)` grabs the event of the timer `DueTime` and the second tick (after `Period` elapses`). Could you do it with `DelaySubscription`? Yes. But you could also do it with `Delay`, `SubscribeWhen`, `Merge`, etc. For this simple case, discussion advantages don't really make that much sense because most of these solutions have little to no overhead. It mostly depends on where you want to take the workflow next and your personal preference.

**Reply by chiyu1203** [reply]
Hi Bruno, I didnt know about that special property of the Timer!! Thanks for the clear explanation!!
